File: src/com/polaris/engine/options/Key.java
/**
 * 
 */
package com.polaris.engine.options;

/**
 * @author lec50
 *
 */
public class Key
{
	
	private static final long KEY_DOUBLE_PRESS = 1000000000 / 5;
	
	private final String keyName;
	private final int keyCode;
	
	private boolean isPressed;
	private boolean wasQuickPressed;
	
	private long pressTimer;
	private boolean isDoublePressed;
	
	public Key(String name, int code)
	{
		keyName = name;
		keyCode = code;
		
		isPressed = false;
		wasQuickPressed = false;
		
		pressTimer = 0;
		isDoublePressed = false;
	}
	
	public final void press()
	{
		isPressed = true;
		wasQuickPressed = false;
		
		long time = System.nanoTime();
		
		if(time - pressTimer < KEY_DOUBLE_PRESS)
			isDoublePressed = true;
		
		pressTimer = time;
	}
	
	public final void release()
	{
		isPressed = false;
		isDoublePressed = false;
		
		if(getPressedTime() <= KEY_DOUBLE_PRESS)
			wasQuickPressed = true;
	}
	
	public final void update()
	{
		if(wasQuickPressed)
			wasQuickPressed = false;
	}
	
	public final String getName()
	{
		return keyName;
	}
	
	public final int getKey()
	{
		return keyCode;
	}
	
	public final boolean isPressed()
	{
		return isPressed;
	}
	
	public final boolean wasQuickPressed()
	{
		return wasQuickPressed;
	}
	
	public final long getPressedTime()
	{
		return System.nanoTime() - pressTimer;
	}
	
	public final boolean isDoublePressed()
	{
		return isDoublePressed;
	}
	
	public String toString()
	{
		return "Key: " + keyName + ", Key Code: " + keyCode + ", isPressed: " + isPressed + ", wasQuickPressed: " + wasQuickPressed + ", \n\t pressTimer: " + pressTimer + ", isDoublePressed: " + isDoublePressed;
	}
}

File: src/polaris/okapi/options/Key.kt
package polaris.okapi.options

import org.lwjgl.glfw.GLFW
import org.lwjgl.glfw.GLFW.glfwGetKeyName
import java.io.Serializable

/**
 * Created by Killian Le Clainche on 12/11/2017.
 */
open class Key(key: Int) : Setting {

    var name: String? = glfwGetKeyName(key, 0)
        private set

    var key: Int = key
        set(value) {
            field = value
            name = glfwGetKeyName(field, 0)
        }

    var isPressed: Boolean = false
        private set

    var isClicked: Boolean = false
        private set

    var wasQuickPressed: Boolean = false

    private var pressTimer: Long = 0

    var isDoublePressed: Boolean = false
        private set

    val pressedTime: Long
        get() = System.nanoTime() - pressTimer

    override fun load(value: String) {
        key = value.toInt()
    }

    override fun save(): String {
        return key.toString()
    }

    fun press() {
        isPressed = true
        wasQuickPressed = false

        val time = System.nanoTime()

        if (time - pressTimer < KEY_DOUBLE_PRESS) isDoublePressed = true

        pressTimer = time
    }

    fun release() {
        isPressed = false
        isDoublePressed = false
        isClicked = true

        if (pressedTime <= KEY_DOUBLE_PRESS) wasQuickPressed = true
    }

    fun update() {
        isClicked = false
        wasQuickPressed = false
    }

    override fun toString(): String {
        return "Key: $name, Key Code: $key, isPressed: $isPressed, wasQuickPressed: $wasQuickPressed, \n\t pressTimer: $pressTimer, isDoublePressed: $isDoublePressed"
    }

    companion object {

        private const val KEY_DOUBLE_PRESS = (1000000000 / 5).toLong()
    }
}